solr/core/src/test/org/apache/solr/cluster/maintenance/InactiveShardRemoverTest.java

```diff
@@ -0,0 +1,231 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.maintenance;
+
+import static org.apache.solr.client.solrj.SolrRequest.METHOD.POST;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.request.CollectionAdminRequest;
+import org.apache.solr.client.solrj.request.V2Request;
+import org.apache.solr.client.solrj.request.beans.PluginMeta;
+import org.apache.solr.client.solrj.response.V2Response;
+import org.apache.solr.cloud.ShardTestUtil;
+import org.apache.solr.cloud.SolrCloudTestCase;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.cloud.DocCollection;
+import org.apache.solr.common.cloud.Slice;
+import org.apache.solr.common.util.TimeSource;
+import org.apache.solr.core.CoreContainer;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class InactiveShardRemoverTest extends SolrCloudTestCase {
+
+  @BeforeClass
+  public static void setupCluster() throws Exception {
+    configureCluster(1)
+        .addConfig(
+            "conf", TEST_PATH().resolve("configsets").resolve("cloud-minimal").resolve("conf"))
+        .configure();
+  }
+
+  @AfterClass
+  public static void tearDownCluster() throws Exception {
+    cluster.shutdown();
+  }
+
+  @Test
+  public void testDeleteInactiveShard() throws Exception {
+
+    addPlugin(new InactiveShardRemoverConfig(1, 0, 2));
+    try {
+      final String collectionName = "testDeleteInactiveShard";
+      createCollection(collectionName, 1);
+
+      final String sliceName =
+          new ArrayList<>(getCollectionState(collectionName).getSlices()).get(0).getName();
+      ShardTestUtil.setSliceState(cluster, collectionName, sliceName, Slice.State.INACTIVE);
+
+      waitForState(
+          "Waiting for inactive shard to be deleted",
+          collectionName,
+          clusterShape(0, 0),
+          5,
+          TimeUnit.SECONDS);
+    } finally {
+      removePlugin();
+    }
+  }
+
+  @Test
+  public void testTtl() throws Exception {
+
+    final int ttlSeconds = 1 + random().nextInt(5);
+    final TimeSource timeSource = cluster.getOpenOverseer().getSolrCloudManager().getTimeSource();
```

Comment:
side comment:  I think TimeSource ought to be a global singleton static thing so it's clear how to get it from absolutely anywhere.  Such singletons can be evil but I think it's warranted for something so universal as tracking time.